src/form/form.js

```diff
@@ -0,0 +1,669 @@
+                return widgets.get(uri).getValue();
+            }
+            return '';
+        },
+
+        /**
+         * Sets the value of a widget
+         * @param {String} uri
+         * @param {String} value
+         * @returns {form}
+         */
+        setValue(uri, value) {
+            if (widgets.has(uri)) {
+                widgets.get(uri).setValue(value);
+            }
+            return this;
+        },
+
+        /**
+         * Gets the values from all the form widgets
+         * @returns {Object}
+         */
+        getValues() {
+            const values = {};
+            for (let [uri, widget] of widgets) {
+                values[uri] = widget.getValue();
+            }
+            return values;
+        },
+
+        /**
+         * Sets the values for the form widgets
+         * @param {Object} values
+         * @returns {form}
+         */
+        setValues(values) {
+            _.forEach(values, (value, uri) => {
+                if (widgets.has(uri)) {
+                    widgets.get(uri).setValue(value);
+                }
+            });
+            return this;
+        },
+
+        /**
+         * Serializes form values to an array of name/value objects
+         * @returns {widgetValue[]}
+         */
+        serialize() {
+            const values = [];
+            for (let widget of widgets.values()) {
+                values.push(widget.serialize());
+            }
+            return values;
+        },
+
+        /**
+         * Validate the form widgets
+         * @returns {Promise}
+         */
+        validate() {
+            const promises = [];
+            for (let [uri, widget] of widgets) {
+                promises.push(
+                    widget.validate()
+                        .catch(messages => Promise.resolve({uri, messages}))
+                );
+            }
+
+            return Promise.all(promises)
+                .then(result => {
+                    let invalid = false;
+
+                    result = _.compact(result);
+
+                    if (result.length) {
+                        result = Promise.reject(result);
+                        invalid = true;
+                    }
+
+                    this.setState('invalid', invalid);
+                    return result;
+                });
+        },
+
+        /**
+         * Submits the form
+         * @returns {form}
+         * @fires submit
+         */
+        submit() {
+            this.validate()
+                .then(() => {
+                    /**
+                     * @event submit
+                     * @param {widgetValue[]} values
+                     */
+                    this.trigger('submit', this.serialize());
+                })
+                .catch(reason => {
+                    /**
+                     * @event invalid
+                     * @param {Object} reason
+                     */
+                    this.trigger('invalid', reason);
```

Comment:
Since the rejection is used to handle user error, it would be good to also support system errors.
If  `reason` is an `Error` (something went wrong within the validation system) we should trigger the `error` event

Reply:
done